### PR TITLE
Add [Project Data] option to case list report

### PR DESCRIPTION
**Pull request by esoergel**
@sravfeyn
http://manage.dimagi.com/default.asp?166981
This might be easiest commit-by-commit.  This PR does a few things:

It explicitly makes a CaseListFilter subclass of the ExpandedMobileWorkerFilter, and pulls out the logic specific to that use-case there.  It's significantly different from the original filter, and this is clearer when each behavior is used.

This moves the build_query method of the case list report to use ESQuery, which is easier to read and reason about.

The CaseListFilter now has no [All Mobile Workers] option anymore, instead it has [Project Data] and [All Data] options.  [All Data] doesn't filter on owner id at all.  [Project Data] filters out stuff we know to be most likely not relevant.  If neither is selected, it falls back to the old behavior of getting an exact set of possible owner_ids and only showing matches.

edit: This is currently on staging, and I just went over it with Sheel to make sure it looks good and does what we want.  That's not exactly QA, but it's something.

### CODE REVIEW

File: corehq/apps/reports/filters/api.py [hunk 1]
```diff
@@ -69,18 +70,27 @@ def get(self, request, domain, all_data=False, share_groups=False):
...
+    def get_location_groups(self):
+        for loc in self.locations_query:
+            group = loc.reporting_group_object()
+            yield (group._id, group.name + ' [group]')
 
+    def get_locations_size(self):
+        return self.locations_query.count()
+
+    def _init_counts(self):
+        groups, _ = self.group_es_call(size=0, return_count=True)
```
Comment (dimagimon) [inline corehq/apps/reports/filters/api.py]: F811 redefinition of unused '_' from line 7

File: corehq/apps/reports/filters/case_list.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    @classmethod
+    def selected_sharing_group_ids(cls, request):
+        emws = request.GET.getlist(cls.slug)
+        return [g[4:] for g in emws if g.startswith("sg__")]
+
+    @classmethod
+    def selected_group_ids(cls, request):
+        return super(CaseListFilter, cls).selected_group_ids(request) + \
+               cls.selected_sharing_group_ids(request)
```
Comment (dimagimon) [inline corehq/apps/reports/filters/case_list.py]: E126 continuation line over-indented for hanging indent

File: corehq/apps/reports/standard/cases/basic.py [hunk 3]
```diff
@@ -156,22 +135,17 @@ def case_owners(self):
         share_group_q = HQESQuery(index="groups").domain(self.domain)\
                                                 .doc_type("Group")\
                                                 .filter(filters.term("case_sharing", True))\
-                                                .filter(filters.term("users", selected_reporting_group_users+selected_user_ids+special_user_ids))\
+                                                .filter(filters.term("users", selected_reporting_group_users+selected_user_ids))\
```
Comment (dimagimon) [inline corehq/apps/reports/standard/cases/basic.py]: E127 continuation line over-indented for visual indent

Comment (dimagimon) [inline corehq/apps/reports/standard/cases/basic.py]: E226 missing whitespace around arithmetic operator

File: corehq/apps/reports/filters/api.py [hunk 4]
```diff
@@ -89,7 +98,7 @@ def get_options(self):
         start = limit * (page - 1)
```
Comment (sravfeyn) [inline corehq/apps/reports/filters/api.py]: A docstring would be helpful to understand what's going on here.

**Comment by czue** [comment]
I haven't reviewed this closely and hard to tell how big of a change it is versus just moving code around. it looks big enough that I feel I should at least ask if this needs QA?

**Comment by esoergel** [comment]
@czue I looked at it with Sheel, and it worked as we expected.  I asked him to take another, more thorough pass, which I'm waiting to hear back on.

**Comment by esoergel** [comment]
Spoke with Sheel about it again today, he gave it a thumbs-up.  I'm confident enough in this to deploy.

**Comment by sravfeyn** [comment]
Cool, will let @czue sign off on the merge.

**Comment by czue** [comment]
sorry i'm a bit backed up but will try and review this tomorrow. don't hesitate to get your new buddy to look if needed

**Comment by esoergel** [comment]
@nickpell would you mind taking a pass at this?  It's a bit of a big one, although it's mostly moving code around.  If you like, I can walk you through it in person.

**Pull request merged**